File: android/app/src/main/java/com/curio/app/data/model/Novel.kt
package com.curio.app.data.model

import com.google.gson.annotations.SerializedName

// ── Novels ─────────────────────────────────────────────────────

data class Novel(
    val id: Long = 0,
    val title: String = "",
    val author: String = "",
    @SerializedName("cover_image_url")
    val coverImageUrl: String = "",
    val description: String = "",
    val source: String = "gutenberg",
    @SerializedName("source_url")
    val sourceUrl: String = "",
    @SerializedName("total_chapters")
    val totalChapters: Int = 0,
    val language: String = "en",
    @SerializedName("category_id")
    val categoryId: Long = 0,
    val likes: Int = 0,
    @SerializedName("created_at")
    val createdAt: String = ""
)

data class NovelListResponse(
    val novels: List<Novel> = emptyList(),
    val total: Long = 0,
    val page: Int = 1,
    val limit: Int = 20
)

data class NovelChapter(
    val id: Long = 0,
    @SerializedName("novel_id")
    val novelId: Long = 0,
    @SerializedName("chapter_number")
    val chapterNumber: Int = 0,
    val title: String = "",
    val body: String = "",
    @SerializedName("read_time_secs")
    val readTimeSecs: Int = 0,
    @SerializedName("created_at")
    val createdAt: String = ""
)

data class NovelChaptersResponse(
    val chapters: List<NovelChapter> = emptyList(),
    val total: Int = 0
)

data class NovelProgress(
    val id: Long = 0,
    @SerializedName("device_id")
    val deviceId: String = "",
    @SerializedName("novel_id")
    val novelId: Long = 0,
    @SerializedName("last_chapter")
    val lastChapter: Int = 1,
    @SerializedName("last_position")
    val lastPosition: Int = 0,
    val completed: Boolean = false,
    val bookmarked: Boolean = false,
    @SerializedName("updated_at")
    val updatedAt: String = ""
)

data class NovelProgressRequest(
    @SerializedName("device_id")
    val deviceId: String,
    @SerializedName("last_chapter")
    val lastChapter: Int? = null,
    @SerializedName("last_position")
    val lastPosition: Int? = null,
    val completed: Boolean? = null,
    val bookmarked: Boolean? = null
)

data class NovelDetailResponse(
    val novel: Novel = Novel(),
    val chapters: List<NovelChapter> = emptyList(),
    val progress: NovelProgress? = null
)

data class LikeResponse(
    val likes: Int = 0
)

File: backend/models/novel.go
package models

import (
	"time"
	"gorm.io/gorm"
)

// Novel represents a long-form book (public domain / uploaded EPUB/PDF).
type Novel struct {
	ID            uint           `json:"id" gorm:"primaryKey"`
	Title         string         `json:"title" gorm:"not null"`
	Author        string         `json:"author" gorm:"not null;default:''"`
	CoverImageURL string         `json:"cover_image_url" gorm:"default:''"`
	Description   string         `json:"description" gorm:"type:text;default:''"`
	Source        string         `json:"source" gorm:"default:'gutenberg'"` // gutenberg | epub | pdf | upload
	SourceURL     string         `json:"source_url" gorm:"default:''"`
	TotalChapters int            `json:"total_chapters" gorm:"not null;default:0"`
	Language      string         `json:"language" gorm:"default:'en'"`
	CategoryID    uint           `json:"category_id"`
	Likes         int            `json:"likes" gorm:"default:0"`
	CreatedAt     time.Time      `json:"created_at"`
	DeletedAt     gorm.DeletedAt `json:"-" gorm:"index"`
}

// NovelChapter stores one chapter of a novel, pre-split and cleaned.
type NovelChapter struct {
	ID            uint      `json:"id" gorm:"primaryKey"`
	NovelID       uint      `json:"novel_id" gorm:"not null;index"`
	ChapterNumber int       `json:"chapter_number" gorm:"not null"`
	Title         string    `json:"title" gorm:"not null;default:''"`
	Body          string    `json:"body" gorm:"type:text;not null"`
	ReadTimeSecs  int       `json:"read_time_secs" gorm:"not null;default:0"`
	CreatedAt     time.Time `json:"created_at"`
}

// UserNovelProgress tracks a user's reading progress per novel (device-based).
type UserNovelProgress struct {
	ID           uint      `json:"id" gorm:"primaryKey"`
	DeviceID     string    `json:"device_id" gorm:"index;not null"`
	NovelID      uint      `json:"novel_id" gorm:"not null"`
	LastChapter  int       `json:"last_chapter" gorm:"not null;default:1"`
	LastPosition int       `json:"last_position" gorm:"not null;default:0"`
	Completed    bool      `json:"completed" gorm:"not null;default:false"`
	Bookmarked   bool      `json:"bookmarked" gorm:"not null;default:false"`
	UpdatedAt    time.Time `json:"updated_at"`
}

// ── Response Types ─────────────────────────────────────────────

type NovelsListResponse struct {
	Novels []Novel `json:"novels"`
	Total  int64   `json:"total"`
	Page   int     `json:"page"`
	Limit  int     `json:"limit"`
}

type NovelDetailResponse struct {
	Novel    Novel         `json:"novel"`
	Chapters []NovelChapter `json:"chapters"`
	Progress *UserNovelProgress `json:"progress,omitempty"`
}

type ProgressUpdateRequest struct {
	DeviceID     string `json:"device_id" binding:"required"`
	LastChapter  int    `json:"last_chapter"`
	LastPosition int    `json:"last_position"`
	Completed    *bool  `json:"completed,omitempty"`
	Bookmarked   *bool  `json:"bookmarked,omitempty"`
}
